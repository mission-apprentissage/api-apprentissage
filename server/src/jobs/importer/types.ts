import type { CronDef } from "job-processor";
import type { ImportStatus } from "shared";

export type Importer = CronDef & {
  getStatus(): Promise<ImportStatus>;
};
