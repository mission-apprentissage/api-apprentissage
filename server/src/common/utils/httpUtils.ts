import axios, { AxiosInstance, AxiosRequestConfig, CreateAxiosDefaults } from "axios";
import http from "http";
import https from "https";
import iconv from "iconv-lite";
import { compose } from "oleoduc";

import logger from "@/common/logger";

// https://github.com/axios/axios/issues/3845#issuecomment-1040819908
class BufferedHttpAgent extends http.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);

    this.highWaterMark = highWaterMark;
  }
  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}
class BufferedHttpsAgent extends https.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);
    //see https://github.com/nodejs/node/issues/39092
    this.highWaterMark = highWaterMark;
  }

  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}

type FetchOptions = AxiosRequestConfig & {
  highWaterMark: number;
  raw: boolean;
  encoding: string;
};
async function _fetch(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { method = "GET", data, highWaterMark, ...rest } = options;
  logger.debug(`${method} ${url}...`);

  return client.request({
    url,
    method,
    httpAgent: new BufferedHttpAgent({ highWaterMark }),
    httpsAgent: new BufferedHttpsAgent({ highWaterMark }),
    ...(data ? { data } : {}),
    ...rest,
  });
}

async function fetchStream(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, encoding = "UTF-8", ...rest } = options;
  const response = await _fetch(url, { responseType: "stream", ...rest }, client);

  const stream = compose(response.data, iconv.decodeStream(encoding));

  if (raw) {
    return {
      headers: response.headers,
      stream,
    };
  } else {
    return stream;
  }
}

async function fetchData(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, ...rest } = options;
  const response = await _fetch(url, { responseType: "json", ...rest }, client);
  const data = response.data;

  if (raw) {
    return {
      headers: response.headers,
      data,
    };
  } else {
    return data;
  }
}

async function fetchJson(url: string, options: Partial<FetchOptions> = {}) {
  const response = await _fetch(url, { ...options, responseType: "json" });
  return response.data;
}

function addCsvHeaders(filename: string, encoding: string, res: any) {
  res.setHeader("Content-disposition", `attachment; filename=${filename}`);
  res.setHeader("Content-Type", `text/csv; charset=${encoding}`);
}

const getHttpClient = (options: CreateAxiosDefaults<any> = {}) =>
  axios.create({
    timeout: 15000,
    httpAgent: new http.Agent({ keepAlive: true }),
    httpsAgent: new https.Agent({ keepAlive: true }),
    ...options,
  });

export { addCsvHeaders, fetchJson, fetchStream, fetchData, getHttpClient };
