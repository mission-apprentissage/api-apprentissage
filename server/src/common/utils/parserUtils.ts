import { Options, parse } from "csv-parse";
import { isEmpty, pickBy } from "lodash-es";

export const DEFAULT_DELIMITER = ";";

export function parseCsv(options: Options = {}) {
  return parse({
    trim: true,
    delimiter: DEFAULT_DELIMITER,
    columns: true,
    bom: true,
    relax_column_count: true,
    on_record: (record) => {
      return pickBy(record, (v) => {
        return !isEmpty(v) && v.trim().length;
      });
    },
    ...options,
  });
}
