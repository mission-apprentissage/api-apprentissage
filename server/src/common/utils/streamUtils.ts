import { Transform } from "node:stream";

type Options = {
  size: number;
};

export function createBatchTransformStream(opts: Options): Transform {
  let currentBatch: unknown[] = [];

  return new Transform({
    objectMode: true,
    transform(chunk, _encoding, callback) {
      currentBatch.push(chunk);

      if (currentBatch.length >= opts.size) {
        this.push(currentBatch);
        currentBatch = [];
      }

      callback();
    },
    flush(callback) {
      if (currentBatch.length > 0) {
        this.push(currentBatch);
      }
      callback();
    },
  });
}
