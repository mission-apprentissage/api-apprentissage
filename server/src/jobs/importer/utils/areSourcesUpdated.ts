export function areSourcesUpdated(
  lastImportSources: Record<string, { import_date: Date }> | null | undefined,
  currentSources: Record<string, { import_date: Date }>
): boolean {
  if (!lastImportSources) {
    return true;
  }

  for (const source in currentSources) {
    if (!lastImportSources[source] || lastImportSources[source].import_date < currentSources[source].import_date) {
      return true;
    }
  }

  return false;
}
