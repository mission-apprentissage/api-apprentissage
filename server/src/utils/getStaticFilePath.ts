import path from "path";
import { __dirname } from "api-alternance-sdk/internal";

export function getStaticDirPath(): string {
  // tsup build project into a single file to dist/index.js
  // Hence import.meta.url references dist/index.js
  if (process.env.IS_BUILT === "true") {
    return path.join(__dirname(import.meta), "../static");
  }

  // When ran directly from source code with tsx or jest
  // Then files are not compiled into a single file
  return path.join(__dirname(import.meta), "../static");
}

export function getStaticFilePath(relativeFilename: string): string {
  return path.join(getStaticDirPath(), relativeFilename);
}
