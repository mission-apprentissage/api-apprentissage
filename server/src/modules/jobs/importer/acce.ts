import type { ReadStream } from "node:fs";
import { Transform } from "node:stream";
import { pipeline } from "node:stream/promises";

import { internal } from "@hapi/boom";
import { parse } from "csv-parse";
import { ObjectId } from "mongodb";
import { IAcce, ZAcceByType } from "shared/models/acce/acce.model";
import unzipper from "unzipper";

import parentLogger from "@/common/logger";

import { downloadCsvExtraction } from "../../../common/apis/acce";
import { withCause } from "../../../common/errors/withCause";
import { getDbCollection } from "../../../common/utils/mongodbUtils";
import { createBatchTransformStream } from "../../../common/utils/streamUtils";

const logger = parentLogger.child({ module: "import:acce" });

async function parseAcceFile(stream: ReadStream, source: string, date: Date) {
  try {
    const zod = ZAcceByType[source as keyof typeof ZAcceByType] ?? null;

    if (zod === null) {
      throw internal("import.acce: unexpected source file", { source });
    }

    await pipeline(
      stream,
      parse({
        bom: true,
        columns: true,
        relax_column_count: true,
        encoding: "latin1",
        delimiter: ";",
        trim: true,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        onRecord: (record, { columns }: any) => {
          const data = columns.reduce((acc: Record<string, string | null>, column: { name: string }) => {
            acc[column.name] = record[column.name]?.trim() || null;
            return acc;
          }, {});

          return zod.parse({
            _id: new ObjectId(),
            source,
            date,
            data,
          });
        },
      }),
      createBatchTransformStream({ size: 1_000 }),
      new Transform({
        objectMode: true,
        async transform(chunk, _encoding, callback) {
          try {
            await getDbCollection("source.acce").insertMany(chunk);
            callback();
          } catch (error) {
            callback(withCause(internal("import.acce: error when inserting"), error));
          }
        },
      })
    );

    await getDbCollection("source.acce").deleteMany({
      source: source as IAcce["source"],
      date: { $ne: date },
    });
  } catch (error) {
    throw withCause(internal("import.acce: unable to parseAcceFile", { source }), error);
  }
}

export async function importAcceData(readStream: ReadStream) {
  const importDate = new Date();

  const zip = readStream.pipe(unzipper.Parse({ forceStream: true }));
  for await (const entry of zip) {
    await parseAcceFile(entry, entry.path, importDate);
    entry.autodrain();
  }
}

export async function runAcceImporter() {
  logger.info("Geting ACCE file...");

  const stream = await downloadCsvExtraction();

  logger.info("Import ACCE data starting...");

  await importAcceData(stream);
}
