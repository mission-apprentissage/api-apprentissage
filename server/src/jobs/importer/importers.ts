import { getAcceeImporterStatus, runAcceImporter } from "./acce/acce.js";
import { getBcnImporterStatus, runBcnImporter } from "./bcn/bcn.importer.js";
import { getCatalogueImporterStatus, runCatalogueImporter } from "./catalogue/catalogue.importer.js";
import { getCertificationImporterStatus, importCertifications } from "./certifications/certifications.importer.js";
import { getCommuneImporterStatus, runCommuneImporter } from "./commune/commune.importer.js";
import { getDaresApiIdccImporterStatus, runDaresApeIdccImporter } from "./dares/ape_idcc/dares.ape_idcc.importer.js";
import { getDaresCcnImporterStatus, runDaresConventionCollectivesImporter } from "./dares/ccn/dares.ccn.importer.js";
import { getFormationsImporterStatus, importFormations } from "./formation/formation.importer.js";
import { getFranceCompetencesImporterStatus, runRncpImporter } from "./france_competence/france_competence.importer.js";
import { getKaliImporterStatus, runKaliConventionCollectivesImporter } from "./kali/kali.ccn.importer.js";
import { getKitApprentissageImporterStatus, runKitApprentissageImporter } from "./kit/kitApprentissage.importer.js";
import { getMissionLocaleImporterStatus, runMissionLocaleImporter } from "./mission_locale/mission_locale.importer.js";
import { getOrganismesImporterStatus, importOrganismes } from "./organisme/organisme.importer.js";
import { getReferentielImporterStatus, runReferentielImporter } from "./referentiel/referentiel.js";
import type { Importer } from "./types.js";

const timings = {
  import_source: "0 4 * * *",
  import_source_main: "0 */4 * * *",
};

export const importers: Record<string, Importer> = {
  "Mise à jour acce": {
    cron_string: timings.import_source,
    handler: runAcceImporter,
    resumable: true,
    getStatus: getAcceeImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données BCN": {
    cron_string: timings.import_source_main,
    handler: runBcnImporter,
    resumable: true,
    getStatus: getBcnImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données Kit Apprentissage": {
    cron_string: timings.import_source_main,
    handler: runKitApprentissageImporter,
    resumable: true,
    getStatus: getKitApprentissageImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données Referentiel": {
    cron_string: timings.import_source_main,
    handler: runReferentielImporter,
    resumable: true,
    getStatus: getReferentielImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données Catalogue": {
    cron_string: timings.import_source_main,
    handler: runCatalogueImporter,
    resumable: true,
    getStatus: getCatalogueImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données France Compétences": {
    cron_string: timings.import_source,
    handler: runRncpImporter,
    resumable: true,
    getStatus: getFranceCompetencesImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des certifications": {
    cron_string: timings.import_source_main,
    handler: async () => importCertifications(),
    resumable: true,
    getStatus: getCertificationImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des organismes": {
    cron_string: timings.import_source_main,
    handler: async () => importOrganismes(false),
    resumable: true,
    getStatus: getOrganismesImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des formations": {
    cron_string: timings.import_source_main,
    handler: importFormations,
    resumable: true,
    getStatus: getFormationsImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Conventions Collective Kali": {
    cron_string: timings.import_source,
    handler: runKaliConventionCollectivesImporter,
    resumable: true,
    getStatus: getKaliImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Conventions Collective Dares": {
    cron_string: timings.import_source,
    handler: runDaresConventionCollectivesImporter,
    resumable: true,
    getStatus: getDaresCcnImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des APE-IDCC Dares": {
    cron_string: timings.import_source,
    handler: runDaresApeIdccImporter,
    resumable: true,
    getStatus: getDaresApiIdccImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Missions Locales": {
    cron_string: timings.import_source,
    handler: runMissionLocaleImporter,
    resumable: true,
    getStatus: getMissionLocaleImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Communes": {
    cron_string: timings.import_source,
    handler: runCommuneImporter,
    resumable: true,
    getStatus: getCommuneImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
};
