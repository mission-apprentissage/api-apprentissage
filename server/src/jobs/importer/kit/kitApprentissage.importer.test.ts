import { useMongo } from "@tests/mongo.test.utils.js";
import { dirname, join } from "path";
import { fileURLToPath } from "url";
import { beforeEach, describe, expect, it, vi } from "vitest";

import { getDbCollection } from "@/services/mongodb/mongodbService.js";
import { getStaticFilePath } from "@/utils/getStaticFilePath.js";

import { runKitApprentissageImporter } from "./kitApprentissage.importer.js";

vi.mock("@/utils/getStaticFilePath", () => ({
  getStaticFilePath: vi.fn(),
}));

describe("runKitApprentissageImporter", () => {
  useMongo();

  beforeEach(async () => {
    vi.useFakeTimers();

    return () => vi.useRealTimers();
  });

  it("should import Kit Apprentissage single_file source", async () => {
    const date = new Date("2023-04-08T22:00:00.000Z");
    vi.setSystemTime(date);

    vi.mocked(getStaticFilePath).mockImplementation((path) =>
      join(dirname(fileURLToPath(import.meta.url)), `fixtures/single_file`, path)
    );

    const result = await runKitApprentissageImporter();

    expect(result).toBe(10);

    const coll = getDbCollection("source.kit_apprentissage");
    const data = await coll.find({}).toArray();
    expect(data.map((datum) => ({ ...datum, _id: "ObjectId" }))).toMatchSnapshot();
  });

  it("should support consecutive import", async () => {
    const date1 = new Date("2023-04-08T22:00:00.000Z");
    vi.setSystemTime(date1);

    vi.mocked(getStaticFilePath).mockImplementation((path) =>
      join(dirname(fileURLToPath(import.meta.url)), `fixtures/single_file`, path)
    );

    const result = await runKitApprentissageImporter();
    expect(result).toBe(10);

    const date2 = new Date("2023-04-09T22:00:00.000Z");
    vi.setSystemTime(date2);

    const result2 = await runKitApprentissageImporter();
    expect(result2).toBe(10);

    const coll = getDbCollection("source.kit_apprentissage");
    const data = await coll.find({ date: date1 }).toArray();
    expect(data).toEqual([]);
  });

  it("should throw an error if importKitApprentissageSource fails", async () => {
    const dataFixture = join(dirname(fileURLToPath(import.meta.url)), `fixtures/non-existing-file.csv`);
    vi.mocked(getStaticFilePath).mockReturnValue(dataFixture);

    await expect(runKitApprentissageImporter()).rejects.toThrowError(
      "import.kit_apprentissage: unable to runKitApprentissageImporter"
    );
  });

  it("should import Kit Apprentissage multiple_files source", async () => {
    const date = new Date("2023-04-08T22:00:00.000Z");
    vi.setSystemTime(date);

    vi.mocked(getStaticFilePath).mockImplementation((path) =>
      join(dirname(fileURLToPath(import.meta.url)), `fixtures/multiple_files`, path)
    );

    const result = await runKitApprentissageImporter();

    expect(result).toBe(50);

    const coll = getDbCollection("source.kit_apprentissage");
    const stats = await coll
      .aggregate([
        {
          $group: {
            _id: { cfd: "$data.Code DiplÃ´me", rncp: "$data.FicheRNCP" },
            sources: {
              $addToSet: "$source",
            },
            count: { $sum: 1 },
          },
        },
        {
          $sort: { "_id.cfd": 1, "_id.rncp": 1 },
        },
        {
          $project: {
            _id: 1,
            count: 1,
            sources: {
              $sortArray: {
                input: "$sources",
                sortBy: 1,
              },
            },
          },
        },
      ])
      .toArray();

    expect(stats).toEqual([
      {
        _id: {
          cfd: "17021006",
          rncp: "RNCP11532",
        },
        count: 31,
        sources: [
          "Kit apprentissage et RNCP v1.0.csv",
          "Kit apprentissage et RNCP v1.1.csv",
          "Kit apprentissage et RNCP v1.2.csv",
          "Kit apprentissage et RNCP v1.3.csv",
          "Kit apprentissage et RNCP v1.4.csv",
          "Kit apprentissage et RNCP v1.5.csv",
          "Kit apprentissage et RNCP v1.6.csv",
          "Kit apprentissage et RNCP v1.7.csv",
          "Kit apprentissage et RNCP v1.8.csv",
          "Kit apprentissage et RNCP v1.9.csv",
          "Kit apprentissage et RNCP v2.0.csv",
          "Kit apprentissage et RNCP v2.1.csv",
          "Kit apprentissage et RNCP v2.2.csv",
          "Kit apprentissage et RNCP v2.3.csv",
          "Kit apprentissage et RNCP v2.4.csv",
          "Kit apprentissage et RNCP v2.5.csv",
          "Kit apprentissage et RNCP v2.6.csv",
          "Kit apprentissage et RNCP v2.7.csv",
          "Kit apprentissage et RNCP v2.8.csv",
          "Kit apprentissage et RNCP v2.9.csv",
          "Kit apprentissage et RNCP v3.0.csv",
          "Kit_apprentissage_20240119.csv",
          "Kit_apprentissage_20240223.csv",
          "Kit_apprentissage_20240329.csv",
        ],
      },
      {
        _id: {
          cfd: "17021006",
          rncp: "RNCP1997",
        },
        count: 3,
        sources: [
          "Kit apprentissage et RNCP v1.0.csv",
          "Kit apprentissage et RNCP v1.1.csv",
          "Kit apprentissage et RNCP v1.2.csv",
        ],
      },
      {
        _id: {
          cfd: "17021006",
          rncp: "RNCP5912",
        },
        count: 3,
        sources: [
          "Kit apprentissage et RNCP v1.0.csv",
          "Kit apprentissage et RNCP v1.1.csv",
          "Kit apprentissage et RNCP v1.2.csv",
        ],
      },
      {
        _id: {
          cfd: "46X33608",
          rncp: "RNCP36906",
        },
        count: 2,
        sources: ["Kit apprentissage et RNCP v2.9.csv", "Kit apprentissage et RNCP v3.0.csv"],
      },
      {
        _id: {
          cfd: "46X33608",
          rncp: "RNCP38447",
        },
        count: 3,
        sources: ["Kit_apprentissage_20240119.csv", "Kit_apprentissage_20240223.csv", "Kit_apprentissage_20240329.csv"],
      },
      {
        _id: {
          cfd: "56T23207",
          rncp: "RNCP25720",
        },
        count: 3,
        sources: [
          "Kit apprentissage et RNCP v1.3.csv",
          "Kit apprentissage et RNCP v1.4.csv",
          "Kit apprentissage et RNCP v1.5.csv",
        ],
      },
      {
        _id: {
          cfd: "56T23207",
          rncp: "RNCP35507",
        },
        count: 5,
        sources: [
          "Kit apprentissage et RNCP v1.6.csv",
          "Kit apprentissage et RNCP v1.7.csv",
          "Kit apprentissage et RNCP v1.8.csv",
          "Kit apprentissage et RNCP v1.9.csv",
          "Kit apprentissage et RNCP v2.0.csv",
        ],
      },
    ]);

    const data = await coll.find({}).toArray();
    expect(data.map((datum) => ({ ...datum, _id: "ObjectId" }))).toMatchSnapshot();
  });
});
