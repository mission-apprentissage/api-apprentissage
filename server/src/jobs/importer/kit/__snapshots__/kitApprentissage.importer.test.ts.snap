// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runKitApprentissageImporter > should import Kit Apprentissage source 1`] = `
[
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": "Licence Professionnelle",
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "oui",
      "Code Diplôme": "25031012",
      "Date d'échéance de la fiche RNCP": "01/01/2025",
      "Dernière MaJ": "v1.0",
      "Fiche ACTIVE/INACTIVE": "ACTIVE",
      "FicheRNCP": "RNCP30111",
      "Intitulé diplôme (DEPP)": "METIERS DE LA QUALITE (LP TOULON)",
      "Niveau fiche RNCP": "Niveau 6",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": "Licence Professionnelle",
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "oui",
      "Code Diplôme": "25024104",
      "Date d'échéance de la fiche RNCP": "01/01/2025",
      "Dernière MaJ": "v1.0",
      "Fiche ACTIVE/INACTIVE": "ACTIVE",
      "FicheRNCP": "RNCP30110",
      "Intitulé diplôme (DEPP)": "METIERS DE LA MODE (LP NIMES)",
      "Niveau fiche RNCP": "Niveau 6",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": null,
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "oui",
      "Code Diplôme": "16X31062",
      "Date d'échéance de la fiche RNCP": "20/07/2027",
      "Dernière MaJ": "V2.8",
      "Fiche ACTIVE/INACTIVE": "ACTIVE",
      "FicheRNCP": "RNCP36741",
      "Intitulé diplôme (DEPP)": "EXPERT EN CONTROLE DE GESTION, AUDIT ET GESTION DE SYSTEME D'INFORMATION (MS SKEMA BUSINESS SCHOOL)",
      "Niveau fiche RNCP": "Niveau 7",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": "Licence Professionnelle",
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "oui",
      "Code Diplôme": "25022158",
      "Date d'échéance de la fiche RNCP": "01/01/2025",
      "Dernière MaJ": "v1.0",
      "Fiche ACTIVE/INACTIVE": "ACTIVE",
      "FicheRNCP": "RNCP30113",
      "Intitulé diplôme (DEPP)": "METIERS DE LA SANTE : NUTRITION, ALIMENTATION (LP RENNES 1)",
      "Niveau fiche RNCP": "Niveau 6",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": "MASTER",
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "oui",
      "Code Diplôme": "13511109",
      "Date d'échéance de la fiche RNCP": "01/01/2025",
      "Dernière MaJ": "v1.0",
      "Fiche ACTIVE/INACTIVE": "ACTIVE",
      "FicheRNCP": "RNCP31803",
      "Intitulé diplôme (DEPP)": "CHIMIE (MASTER GRENOBLE ALPES)",
      "Niveau fiche RNCP": "Niveau 7",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": "Licence Professionnelle",
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "oui",
      "Code Diplôme": "25022204",
      "Date d'échéance de la fiche RNCP": "01/01/2025",
      "Dernière MaJ": "v1.0",
      "Fiche ACTIVE/INACTIVE": "ACTIVE",
      "FicheRNCP": "RNCP30050",
      "Intitulé diplôme (DEPP)": "CHIMIE : FORMULATION (LP TOULOUSE 3)",
      "Niveau fiche RNCP": "Niveau 6",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": "Titre ingénieur",
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "oui",
      "Code Diplôme": "17021004",
      "Date d'échéance de la fiche RNCP": "01/01/2024",
      "Dernière MaJ": "v1.0",
      "Fiche ACTIVE/INACTIVE": "INACTIVE",
      "FicheRNCP": "RNCP29277",
      "Intitulé diplôme (DEPP)": "INGENIEUR DIPLOME DE L'INSTITUT SUPERIEUR D'AGRICULTURE YNCREA HAUTS-DE-FRANCE",
      "Niveau fiche RNCP": "Niveau 7",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": "Titre ingénieur",
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "non",
      "Code Diplôme": "17032620",
      "Date d'échéance de la fiche RNCP": "31/08/2026",
      "Dernière MaJ": "v1.2",
      "Fiche ACTIVE/INACTIVE": "ACTIVE",
      "FicheRNCP": "RNCP36416",
      "Intitulé diplôme (DEPP)": "INGENIEUR DIPLOME DE L'ECOLE SUPERIEURE D'INGENIEURS DE RENNES DE L'UNIVERSITE RENNES 1, SPECIALITE TECHNOLOGIES DE L'INFORMATION",
      "Niveau fiche RNCP": "Niveau 7",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": "Licence Professionnelle",
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "oui",
      "Code Diplôme": "25022154",
      "Date d'échéance de la fiche RNCP": "01/01/2025",
      "Dernière MaJ": "v1.0",
      "Fiche ACTIVE/INACTIVE": "ACTIVE",
      "FicheRNCP": "RNCP30074",
      "Intitulé diplôme (DEPP)": "INDUSTRIES AGROALIMENTAIRES : GESTION, PRODUCTION ET VALORISATION (LP CNAM)",
      "Niveau fiche RNCP": "Niveau 6",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
  {
    "_id": "ObjectId",
    "data": {
      "Abrégé de diplôme (RNCP)": "DGE_GM",
      "Accès à l'apprentissage de la fiche RNCP (oui/non)": "oui",
      "Code Diplôme": "15531005",
      "Date d'échéance de la fiche RNCP": "31/08/2019",
      "Dernière MaJ": "v1.0",
      "Fiche ACTIVE/INACTIVE": "INACTIVE",
      "FicheRNCP": "RNCP27423",
      "Intitulé diplôme (DEPP)": "DIPLOME GRADE MASTER DE PARIS DAUPHINE : MANAGEMENT INTERNATIONAL",
      "Niveau fiche RNCP": "Niveau 7",
    },
    "date": 2023-04-08T22:00:00.000Z,
    "source": "kit_apprentissage_20240119.csv",
  },
]
`;
